File: backend/app/spec/arch_spec_v1.py
from typing import Dict, List, Literal, Optional, Any

from pydantic import BaseModel, Field


class ProjectMeta(BaseModel):
    name: str
    type: Optional[str] = None
    description: Optional[str] = None


class Resources(BaseModel):
    cpu: Optional[str] = None
    memory: Optional[str] = None


class Security(BaseModel):
    tls: Optional[bool] = None
    notes: Optional[str] = None


class Storage(BaseModel):
    type: Optional[str] = None  # e.g. postgres, s3, local
    size: Optional[str] = None  # e.g. 10Gi
    path: Optional[str] = None


class Node(BaseModel):
    id: str
    kind: str
    name: str
    runtime: Optional[str] = None  # e.g. fastapi, node, nextjs
    ports: List[int] = Field(default_factory=list)
    env: Dict[str, str] = Field(default_factory=dict)
    resources: Optional[Resources] = None
    security: Optional[Security] = None
    storage: Optional[Storage] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class Edge(BaseModel):
    id: str
    source: str
    target: str
    protocol: Optional[str] = "http"
    direction: Optional[Literal["uni", "bi"]] = "uni"
    security: Optional[Security] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class Deploy(BaseModel):
    target: Literal["compose", "k8s"] = "compose"
    gateway: bool = False
    dbMode: Optional[str] = None  # local | cloud | mixed
    outputName: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ProjectSpec(BaseModel):
    version: Literal["v1"] = "v1"
    project: ProjectMeta
    nodes: List[Node]
    edges: List[Edge]
    deploy: Deploy

File: frontend/auto_arch/src/spec/archSpecV1.ts
import { Edge as RFEdge, Node as RFNode } from 'reactflow';

export type DeployTarget = 'compose' | 'k8s';

export interface ProjectMeta {
  name: string;
  type?: string;
  description?: string;
}

export interface Resources {
  cpu?: string;
  memory?: string;
}

export interface Security {
  tls?: boolean;
  notes?: string;
}

export interface Storage {
  type?: string;
  size?: string;
  path?: string;
}

export interface NodeSpec {
  id: string;
  kind: string;
  name: string;
  runtime?: string;
  ports: number[];
  env: Record<string, string>;
  resources?: Resources;
  security?: Security;
  storage?: Storage;
  metadata: Record<string, unknown>;
}

export interface EdgeSpec {
  id: string;
  source: string;
  target: string;
  protocol?: string;
  direction?: 'uni' | 'bi';
  security?: Security;
  metadata: Record<string, unknown>;
}

export interface DeploySpec {
  target: DeployTarget;
  gateway: boolean;
  dbMode?: string;
  outputName?: string;
  metadata: Record<string, unknown>;
}

export interface ProjectSpec {
  version: 'v1';
  project: ProjectMeta;
  nodes: NodeSpec[];
  edges: EdgeSpec[];
  deploy: DeploySpec;
}

export interface ConvertOptions {
  projectName?: string;
  projectType?: string;
  deployTarget?: DeployTarget;
  gateway?: boolean;
  dbMode?: string;
  outputName?: string;
}

const toNumberArray = (value: unknown): number[] => {
  if (value === undefined || value === null) return [];
  if (Array.isArray(value)) {
    return value
      .map((v) => Number(v))
      .filter((v) => !Number.isNaN(v));
  }
  const num = Number(value);
  return Number.isNaN(num) ? [] : [num];
};

export const reactFlowToSpecV1 = (
  nodes: RFNode[],
  edges: RFEdge[],
  options: ConvertOptions = {}
): ProjectSpec => {
  const nodeSpecs: NodeSpec[] = nodes.map((node) => {
    const data = node.data || {};
    const ports = toNumberArray((data as any).port || (data as any).ports);

    return {
      id: node.id,
      kind: String(data.type || node.type || 'component'),
      name: String(data.label || node.id),
      runtime: (data as any).runtime,
      ports,
      env: (data as any).env || {},
      resources: (data as any).resources,
      security: (data as any).security || ((data as any).tls ? { tls: true } : undefined),
      storage: (data as any).storage,
      metadata: {
        ...data,
        position: node.position,
      },
    };
  });

  const edgeSpecs: EdgeSpec[] = edges.map((edge) => {
    const data = edge.data || {};
    return {
      id: edge.id,
      source: edge.source,
      target: edge.target,
      protocol: (data as any).protocol || 'http',
      direction: ((data as any).direction === 'bi' ? 'bi' : 'uni') as 'uni' | 'bi',
      security: (data as any).tls ? { tls: true } : undefined,
      metadata: data as Record<string, unknown>,
    };
  });

  const deploy: DeploySpec = {
    target: options.deployTarget || 'compose',
    gateway: options.gateway ?? true,
    dbMode: options.dbMode,
    outputName: options.outputName || options.projectName || 'generated_project',
    metadata: {},
  };

  return {
    version: 'v1',
    project: {
      name: options.projectName || 'untitled-project',
      type: options.projectType,
    },
    nodes: nodeSpecs,
    edges: edgeSpecs,
    deploy,
  };
};
